db/listener_test.cc

```diff
@@ -378,18 +378,35 @@ TEST_F(EventListenerTest, MultiCF) {
     ASSERT_OK(Put(5, "nikitich", std::string(90000, 'n')));
     ASSERT_OK(Put(6, "alyosha", std::string(90000, 'a')));
     ASSERT_OK(Put(7, "popovich", std::string(90000, 'p')));
+
+    bool call_back = false;
     for (int i = 1; i < 8; ++i) {
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "DBImpl::NotifyOnFlushCompleted::PostAllOnFlushCompleted",
+          [&](void* /*arg*/) {
+            ASSERT_EQ(listener->flushed_dbs_.size(), i);
+            ASSERT_EQ(listener->flushed_column_family_names_.size(), i);
+            if (i == 7) {
+              // make sure callback functions are called in the right order
+              for (size_t j = 0; j < cf_names.size(); j++) {
+                ASSERT_EQ(listener->flushed_dbs_[j], db_);
+                ASSERT_EQ(listener->flushed_column_family_names_[j],
+                          cf_names[j]);
+              }
+            }
+            call_back = true;
+          });
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+
       ASSERT_OK(Flush(i));
       ASSERT_OK(dbfull()->TEST_WaitForFlushMemTable());
-      ASSERT_EQ(listener->flushed_dbs_.size(), i);
-      ASSERT_EQ(listener->flushed_column_family_names_.size(), i);
+      ASSERT_TRUE(call_back);
```

Comment:
I would have speculated the problem is TEST_WaitForFlushMemTable does not wait for event listener to finish. What did you think is the problem?

Try this patch, for example:

```
diff --git a/db/db_impl/db_impl_compaction_flush.cc b/db/db_impl/db_impl_compaction_flush.cc
index 4ff87c1e4..8afb1c782 100644
--- a/db/db_impl/db_impl_compaction_flush.cc
+++ b/db/db_impl/db_impl_compaction_flush.cc
@@ -855,6 +855,8 @@ void DBImpl::NotifyOnFlushCompleted(
        mutable_cf_options.level0_stop_writes_trigger);
   // release lock while notifying events
   mutex_.Unlock();
+  bg_cv_.SignalAll();
+  usleep(100000);
   {
     for (auto& info : *flush_jobs_info) {
       info->triggered_writes_slowdown = triggered_writes_slowdown;
```

Reply:
Yeah that's the problem but I am thinking of fixing it in the way we write test. The reason is  the design of "// release lock while notifying events" seems to be optimizing for performance and `TEST_WaitForFlushMemTable`'s underlying `bg_cv_` seems to be waiting for that same lock. 

(Intended to write more context in the PR but I got distracted by my snack ... )


I




(Intended to write more context in the PR summary after getting done with my snack ... which never finished)

Reply:
I think it is close to fixing in the test as you said, but the reason I commented this line in particular is because `call_back` might not have been set to true by the time this assertion executes, because `TEST_WaitForFlushMemTable()` doesn't wait long enough.

Reply:
> I think it is close to fixing in the test as you said, but the reason I commented this line in particular is because `call_back` might not have been set to true by the time this assertion executes, because `TEST_WaitForFlushMemTable()` doesn't wait long enough.

Oh ..... yep I somehow overlooked that and repeat myself with the same mistake behind the flaky test :p